lib/active_merchant/billing/gateways/card_connect.rb

```diff
@@ -0,0 +1,283 @@
+          post[:postal] = address[:zip] if address[:zip]
+          post[:phone] = address[:phone] if address[:phone]
+        end
+      end
+
+      def add_money(post, money)
+        post[:amount] = amount(money)
+      end
+
+      def add_currency(post, money, options)
+        post[:currency] = (options[:currency] || currency(money))
+      end
+
+      def add_invoice(post, options)
+        post[:orderid] = options[:order_id]
+        post[:ecomind] = (options[:recurring] ? 'R' : 'E')
+      end
+
+      def add_payment(post, payment)
+        post[:name] = payment.name
+        if card_brand(payment) == 'check'
+          add_echeck(post, payment)
+        else
+          post[:account] = payment.number
+          post[:expiry] = expdate(payment)
+          post[:cvv2] = payment.verification_value
+        end
+      end
+
+      def add_echeck(post, payment)
+        post[:accttype] = 'ECHK'
+        post[:account] = payment.account_number
+        post[:bankaba] = payment.routing_number
+      end
+
+      def add_reference(post, authorization)
+        post[:retref] = authorization
+      end
+
+      def add_additional_data(post, options)
+        post[:ponumber] = options[:po_number]
+        post[:taxamnt] = options[:tax_amount] if options[:tax_amount]
+        post[:frtamnt] = options[:freight_amount] if options[:freight_amount]
+        post[:dutyamnt] = options[:duty_amount] if options[:duty_amount]
+        post[:orderdate] = options[:order_date] if options[:order_date]
+        post[:shipfromzip] = options[:ship_from_zip] if options[:ship_from_zip]
+        if (shipping_address = options[:shipping_address])
+          post[:shiptozip] = shipping_address[:zip]
+          post[:shiptocountry] = shipping_address[:country]
+        end
+        if options[:items]
+          post[:items] = []
+          options[:items].each do |item|
+            updated = {}
+            item.each_pair do |k, v|
+              updated.merge!(k.to_s.gsub(/_/, '') => v)
+            end
+            post[:items] << updated
+          end
+        end
+      end
+
+      def add_3DS(post, options)
+        post[:secureflag] = options[:secure_flag] if options[:secure_flag]
+        post[:securevalue] = options[:secure_value] if options[:secure_value]
+        post[:securexid] = options[:secure_xid] if options[:secure_xid]
+      end
+
+      def headers
+        {
+          'Authorization' => 'Basic ' + Base64.strict_encode64("#{@options[:username]}:#{@options[:password]}").strip,
+          'Content-Type' => 'application/json'
+        }
+      end
+
+      def expdate(credit_card)
+        "#{format(credit_card.month, :two_digits)}#{format(credit_card.year, :two_digits)}"
+      end
+
+      def parse(body)
+        JSON.parse(body)
+      end
+
+      def url(action)
+        if test?
+          test_url + action
+        else
+          (@options[:domain] ? @options[:domain] : live_url) + action
```

Comment:
This needs to be vetted in some way or we're allowing ANY URL to go through here. You could exfiltrate all our cards by just passing in `postbin.com` or something.

Reply:
(BTW, assuming this is from our convo in Slack, I was suggesting doing a regex vet. So if the URLs always have the form `*.foo.com`, we'd just throw an exception if the domain didn't match that.)

Reply:
Would the test for this be better suited as a unit test?

Reply:
Wouldn't that still allow the customer at runtime to submit any URL?

Reply:
So your saying it should be on the core side? I was going to check in the initialize step before a commit was made.

Reply:
No, I was saying you want it somewhere in the actual AM adapter. Putting it in `initialize` (as opposed to right here in `url`) would be fine, and might provide better error reporting. But I don't think we want this check in Core; while that'd help _us_ out, it both puts more logic into Core, and it doesn't help out other users of AM.